src/IROperator.cpp

```diff
@@ -970,4 +971,31 @@ Tuple tuple_select(const Expr &condition, const Tuple &true_value, const Tuple &
     return result;
 }
 
+// TODO: It would be nice to make these pure intrisics, but I don't
+// think that is actually correct if the thing they're wrapped around
+// is not pure.
+Expr unsafe_promise(Expr condition, Expr value) {
+    user_assert(condition.defined()) << "unsafe_promise with undefined condition.\n";
+    user_assert(condition.type().is_bool()) << "unsafe_promise condition must be a boolean type.\n";
+    user_assert(value.defined()) << "unsafe_promise with undefined value.\n";
+
+    return Internal::Call::make(value.type(),
+                                Internal::Call::unsafe_promise,
+                                {condition, value},
+                                Internal::Call::Intrinsic);
+}
+
+Expr unsafe_promise_clamped(Expr value, Expr min, Expr max) {
```

Comment:
IMHO we'd be better off requiring min and max to be explicitly defined in all cases; if you're inserting known-unsafe code, you should make the contract as explicit as possible.

Also: do we need to check here (or elsewhere) that the value/min/max types are all identical types?

Reply:
In the case where the bound is undefined, it doesn't exist. Hence I would consider this to fail safely if one did not mean it to be undefined. The failure will also likely be obvious as usually one is adding these to satisfy bounds inference requirements and they will not be satisfied if the value is undefined.

The types of min and max are required to losslessly convert to the type of value and are converted to the type of value. This is how it must be to handle constants in the argument slots and is fully safe.